Tools/nxworkspace.1
.\"nxworkspace(1) man page
.\"put together by onflapp
.\"Copyright (C) 2020 Free Software Foundation, Inc.
.\"Copying and distribution of this file, with or without modification,
.\"are permitted in any medium without royalty provided the copyright
.\"notice and this notice are preserved.
.\"
.\"Process this file with
.\"groff -man -Tascii nxworkspace.1
.\"
.TH NXWORKSPACE 1 "August 2020" GNUstep "NEXTSPACE System Manual"
.SH NAME
nxworkspace \- invoke Workspace commands from command line
.SH SYNOPSIS
.B nxworkspace [--open <path> <app>] [--activate <app>]
.P
.SH DESCRIPTION
.P
nxworkspace enables you to use some of Workspace's functionality from directly command line. 

The most common use of this utility is to open a path in Workspace. Workspace will take appropriate action based on what kind of path this actuall is.

If path is a regular file, Workspace will open it in default application.

  nxworkspace --open ~/Documents/mytext.txt

If path is a directory, Workspace will open it in File Viewer

  nxworkspace --open .

If path is an application, Workspace will launch it

  nxworkspace --open /Applications/Terminal.app

You can specify which application Workspace should open path with by passing application name.

  nxworkspace --open ./nxworkspace.1 ProjectCenter

.SH OPTIONS
.P
--open <path>       open path in Workspace

--open <path> <app> open path with specific application

--activate <app>    launch or activate application
.P


NEWS
1 NEWS
******

The currently released version is '2.9.2'.

1.1 Changes in version '2.9.2'
==============================

   * Update for newer versions of Java

   * Various minor bugfixes.

1.2 Changes in version '2.9.1'
==============================

   * Add workaround for Clang bug on Windows MSVC when tests contain no
     Objective-C constructs.

   * Various fixes for minor issues introduced in the test framework by
     parallelisation mof testcase execution.

1.3 Changes in version '2.9.0'
==============================

   * Better check for objc runtime on Windows.

   * Split linker flags to better support partial linking: 'ALL_LDFLAGS'
     is now a combination of 'FINAL_LDFLAGS' and 'ALL_LDFLAGS'.

   * Better support for newer gcc versions.

   * Add support for storyboard files.

   * Increase autoconf version to 2.65 and make autoconf handle
     Objective-C++ and OBJCXX variables directly.

   * Fix bug that prevented ARC from getting used.

   * Link subproject object files directly instead of first merging them
     into 'subproject.o'.

   * Support building on Windows with Clang MSVC target.

   * Improve mingw64 support: for instance, adopt the triplet used by
     the mingw-w64 project rather than using the one returned by
     autoconf.  Fixes building Gorm.

1.4 Changes in version '2.8.0'
==============================

   * Better library combo and ABI detection: gnustep-make will now
     attempt to detect the optimal library combo and ABI supported by
     the installed compiler and Objective-C runtime library.  Explicit
     selection of the runtime ABI is now possible using the
     '--with-runtime-abi' configure option.

   * Full support for the gnustep-2.0 Objective-C ABI.

   * Support for creating a Git tag and creating a tarball from a git
     tag using the 'git-tag' and 'git-dist' targets.

   * Support for creating a Mercurial tag and creating a tarball from a
     hg tag using the 'hg-tag' and 'hg-dist' targets.

   * Dropped legacy Rhapsody and FreeBSD-out support.

1.5 Changes in version '2.7.0'
==============================

Garbage collection support removed

   ARC enabled by default if the ng runtime is used

   Multi-architecture directory layout adjusted to be more Debian
compatible

   Framework support fixes

   Various other minor bugfixes

1.6 Changes in version '2.6.8'
==============================

Configure option '-with-library-combo=ng-gnu-gnu' to use the 'Next
Generation' setup of the latest ObjectiveC-2 runtime and compiler
features rather than traditional runtime.  Requires the new runtime and
a recent clang compiler.

   With the 'ng' runtime in use, you can define GS_WITH_ARC=1 at the
start of a makefile, or in your environment, or in the command line
arguments to have objC code built using ARC.

   Command line option 'documentation=no' to suppress builds of
documentation.

   Integration of testsuite for regression/unit testing of libraries
using the 'check' target.  In your makefile define libraryname_TEST_DIR
= TestsSubdirectory

   Various minor bugfixes, documentation spelling corrections etc.

   The '-enable-strict-v2-mode' option is now, after eight years, turned
on by default (in anticipation of finally removing backward
compatibility with version one).  WARNING; Packagers please ensure that
you update any old gnustep-make version one makefiles.

   Garbage collection support to be removed at the next release.

1.7 Changes in version '2.6.7'
==============================

Improved package building support

   Improved environment variable support

   Improved Java support

   Various minor bugfixes, documentation spelling corrections etc.

1.8 Changes in version '2.6.6'
==============================

Debian packaging support

   Minor bugfixes.

1.9 Changes in version '2.6.5'
==============================

Bugfix for non-fragile ABI test

   Bugfix order if include diorectories for system includes

   Bugfix equality test for prorocol testing

   Added minimal test support for .c and .cc files.

1.10 Changes in version '2.6.4'
===============================

Test framework extended equality tests.

   Android build target

1.11 Changes in version '2.6.3'
===============================

Minor bugfix release.

1.12 Changes in version '2.6.2'
===============================

Added standalone filesystem layout for putting everything in one
directory for easy deployment of relocatable.  Other bug fixes.

1.13 Changes in version '2.6.1'
===============================

Bug fix release.  Most notably to fix a problem compiling GNUstep with
clang.

1.14 Changes in version '2.6.0'
===============================

   * The default filesystem layout is now the 'fhs' layout
   * The default location of the configuration file changed when not
     installing system-wide
   * Removed the -with-system-root, -with-local-root and
     -with-network-root options
   * Removed obsolete variables (deprecated 4 years ago).
   * A new test framework is included in the pacakge
   * objc.make deprecated.  Use tool.make instead
   * -enable-absolute-install-paths is now the default on Darwin

1.15 Changes in version '2.4.0'
===============================

   * Add -enable-objc-nonfragile-abi flag
   * New serial-subdirectories.make and parallel-subdirectories.make.
   * Support for parallel building of subdirectories.
   * Support for parallel building of different instances.
   * Support source files in subdirectories without using a subproject.
   * Support header files in sub-subdirectories.
   * Batch-compile Java files (much faster).
   * Changed default compilation flags to -O2 -g.
   * Fixes for building documentation.
   * Renamed info files to prevent conflicts.
   * Updated cygwin support.

1.16 Changes in version '2.2.0'
===============================

   * New -with-objc-lib-flag for specifying libobjc.
   * Automatically uses -lobjc_gc with configured with garbage
     collection

Tools/nxworkspace.m
/* Copyright (C) 2020 Free Software Foundation, Inc.

   Written by: onflapp
   Created: September 2020

   This file is part of the NEXTSPACE Project

   This program is free software; you can redistribute it and/or
   modify it under the terms of the GNU General Public License
   as published by the Free Software Foundation; either
   version 2 of the License, or (at your option) any later version.

   You should have received a copy of the GNU General Public
   License along with this program; see the file COPYING.
   If not, write to the Free Software Foundation,
   51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.

   */

#import	<AppKit/AppKit.h>

NSString* absolutePath(NSString* path) {
  if ([path isAbsolutePath]) return [path stringByStandardizingPath];
  else {
    NSFileManager* fm = [NSFileManager defaultManager];
    path = [[fm currentDirectoryPath] stringByAppendingPathComponent:path];
    return [path stringByStandardizingPath];
  }
}

void printUsage() {
  fprintf(stderr, "Usage: nxworkspace [--open <path> <app>] [--activate <app>]\n");
  fprintf(stderr, "\n");
  fprintf(stderr, "Help: invoke Workspace commands from command line\n");
  fprintf(stderr, "Options:\n");
  fprintf(stderr, "  --open <path>       open path in Workspace\n");
  fprintf(stderr, "  --open <path> <app> open path with specific application\n");
  fprintf(stderr, "  --activate <app>    launch or activate application\n");
  fprintf(stderr, "\n");
}

int main(int argc, char** argv, char** env)
{
  NSProcessInfo *pInfo;
  NSArray *arguments;
  CREATE_AUTORELEASE_POOL(pool);

#ifdef GS_PASS_ARGUMENTS
  [NSProcessInfo initializeWithArguments:argv count:argc environment:env_c];
#endif

  pInfo = [NSProcessInfo processInfo];
  arguments = [pInfo arguments];

  @try {
    if ([arguments count] == 1) {
      printUsage();
      exit(1);
    }
    else if ([[arguments objectAtIndex:1] isEqualToString:@"--open"] && [arguments count] >= 3 ) {
      NSWorkspace* ws = [NSWorkspace sharedWorkspace];
      NSString* path = absolutePath([arguments objectAtIndex:2]);
      NSString* app = ([arguments count] == 4)?[arguments objectAtIndex:3]:nil;

      [ws openFile:path withApplication:app];
    }
    else if ([[arguments objectAtIndex:1] isEqualToString:@"--activate"] && [arguments count] == 3) {
      NSWorkspace* ws = [NSWorkspace sharedWorkspace];
      NSString* app = [arguments objectAtIndex:2];

      [ws launchApplication:app];
    }
    else {
      printUsage();
      exit(1);
    }
  }
  @catch (NSException* ex) {
    NSLog(@"exception: %@", ex);
    printUsage();
    exit(6);
  }

  RELEASE(pool);

  exit(EXIT_SUCCESS);
}

